### ISSUE TITLE
API for using drawElements rather than drawArrays

### ISSUE THREAD

**Issue by backspaces** [mrdoob/three.js#10666]
(*** This section is for bug reports and feature requests only. This is NOT a help site. Do not ask help questions here. If you need help, please use stackoverflow. ***)

##### Description of the problem 
It seems like the standard geometries have faces as indices/elements. But it appears like gl.drawArrays is being used. I'm assuming that there is a way for drawElements to be used instead, to minimize the attribute buffer sizes, using the face element/index values.

Also, as I get into ShaderMaterials, I'd prefer to have attribute arrays the size of the number of unique vertices, with color, positions etc the same size as well, using drawElements to manage duplications .. i.e. 4 vertices per rectangle, with 6 elements for its two triangles.

What is the method used to have WebGLRenderer use drawElements? Possibly WebGLIndexedBufferRenderer?

##### Three.js version

- [ ] Dev
- [x] r84
- [ ] ...

##### Browser

- [x] All of them
- [ ] Chrome
- [ ] Firefox
- [ ] Internet Explorer

##### OS

- [x] All of them
- [ ] Windows
- [ ] Linux
- [ ] Android
- [ ] IOS

**Comment by WestLangley** [comment]
Use _indexed_ `BufferGeometry`, instead of non-indexed `BufferGeometry`.

An example of its use can be found in `PlaneBufferGeometry`.

**Comment by backspaces** [comment]
Whew, thanks. I didn't find it in the docs, searching for Buffer (nice docs search feature, btw) I found:
```
BufferAttribute
BufferGeometry
InstancedBufferAttribute
InstancedBufferGeometry
InstancedInterleavedBuffer
InterleavedBuffer
InterleavedBufferAttribute
```
Not one of these, right? Or Instanced === Indexed? I'm presuming Instanced refers to http://blog.tojicode.com/2013/07/webgl-instancing-with.html

So maybe this.setIndex( indices ) in BufferGeometry is the difference? That would be superb. And yes, I missed that.

PS: I'm *definitely* not complaining .. I love Three's architecture. And I really understand docs lagging behind implementation, I sure do that. But it's tough on us noobs to figure out how to do things.

But if this issue is a documentation issue, I'd love two things:
* In the severally sparse doc pages, flesh them out with some explanation of how they are used and point to the already existing demos. An example is InstancedBufferGeometry or InstancedBufferAttribute
* Have some *really* dead simple examples of the more advanced features. ShaderMaterial: just one triangle dynamically wandering around a scene would be great, really. And so on with the rest. I love the sexy stuff, but for a developer, the sexy gets in the way of the simple.

**Comment by looeee** [comment]
>And I really understand docs lagging behind implementation...

I went over all the docs a while back and as far as I can tell everything in the API is documented (unless I missed something), although as you mentioned a lot of it is very sparse - certainly the various buffer attributes and geometries could do with being expanded, and perhaps having a manual page added to explain their use. 

>Have some really dead simple examples of the more advanced features.

I think this would be great! The place for these is inline in the documentation - see for example the [MeshStandardMaterial(https://threejs.org/docs/#Reference/Materials/MeshStandardMaterial). More simple inline examples like this in the documentation would definitely help people.

**Comment by mrdoob** [comment]
`Geometry` gets converted to non-indexed `BufferGeometry` at render time. It's like this because this way we can *easily* update the `BufferGeometry` is the user updates the `Geometry`.

If you want to use indexed geometries you'll have to construct them directly using `BufferGeometry` and the `index` property (be aware that we may eventually do a `IndexedBufferGeometry` for clarity).

And no... Instancing is not the same as indexed.

**Comment by backspaces** [comment]
Thanks, *very* helpful. Now I need to know how Three builds shaders for the BufferGeometry or any given mesh. Looking at the BufferGeometry doc, I think position, normal, color, and index are built-in attributes that if I use, I'll not need my own ShaderMaterial .. that Three provides/constructs a shader. 

Not sure about translations & rotations, possibly the positions are pre-multiplied by the appropriate matrix?

Is there any way to see the shaders that Three creates for a given mesh? Maybe a breakpoint at WebGLShader?

Basically I'm trying to see how to use BufferGeometry without writing my own shader. I want to write shaders but only after really understanding the standard Three architecture.

**Comment by mrdoob** [comment]
> Is there any way to see the shaders that Three creates for a given mesh? Maybe a breakpoint at WebGLShader?

https://github.com/spite/ShaderEditorExtension

> Basically I'm trying to see how to use BufferGeometry without writing my own shader. I want to write shaders but only after really understanding the standard Three architecture.

This should be a good example: https://threejs.org/examples/#webgl_buffergeometry_uint

**Comment by backspaces** [comment]
I'm closing this .. I think the answer is basically use a BufferGeometry setting the index property. And likely more importantly, although drawElements has its place, instancing will more likely be what I want. Thanks!
